File: leetcode/js/sum-of-values-at-indices-with-k-set-bits.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var sumIndicesWithKSetBits = function (nums, k) {
  return nums
    .filter((_, idx) => count1Bit(idx) === k)
    .reduce((acc, cur) => acc + cur, 0);
};

function count1Bit(num) {
  return num.toString(2).replaceAll("0", "").length;
}

var sumIndicesWithKSetBits = function (nums, k) {
  let result = 0;

  for (let i = 0; i < nums.length; i++) {
    let count = 0;
    let idx = i;

    while (idx > 0) {
      count += idx & 1;
      idx >>= 1;
    }

    if (count === k) {
      result += nums[i];
    }
  }

  return result;
};

File: leetcode/rust/sum-of-values-at-indices-with-k-set-bits.rs
impl Solution1 {
    pub fn sum_indices_with_k_set_bits(nums: Vec<i32>, k: i32) -> i32 {
        let mut result = 0;

        for i in 0..nums.len() {
            let mut count = 0;
            let mut idx = i;

            while idx > 0 {
                count += idx & 1;
                idx >>= 1;
            }

            if count == (k as usize) {
                result += nums[i];
            }
        }

        result
    }
}

impl Solution2 {
    pub fn sum_indices_with_k_set_bits(nums: Vec<i32>, k: i32) -> i32 {
        nums.iter().enumerate().fold(0, |acc, (idx, num)| {
            if idx.count_ones() == (k as u32) {
                acc + num
            } else {
                acc
            }
        })
    }
}
